sysboundary/ionosphere.cpp

```diff
@@ -3315,9 +3326,64 @@ namespace SBC {
          mpiGrid[cellID]->adjustSingleCellVelocityBlocks(popID,true);
          // TODO: The moments can also be analytically calculated from ionosphere parameters.
          // Maybe that's faster?
-         calculateCellMoments(mpiGrid[cellID], true, false, true);
+         if(popID == getObjectWrapper().particleSpecies.size() - 1) { // we only recompute moments once when all pops are done
+            if(mpiGrid[cellID]->sysBoundaryLayer == 1 || boundaryVDFmode == CopyAndLosscone) { // got new VDFs, the other layers are done manually
+               calculateCellMoments(mpiGrid[cellID], true, false, true);
+            } else {
+               mpiGrid[cellID]->parameters[CellParams::RHOM] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::RHOQ] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VX] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VY] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VZ] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_11] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_22] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_33] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::RHOM_R] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::RHOQ_R] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VX_R] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VY_R] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VZ_R] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_11_R] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_22_R] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_33_R] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::RHOM_V] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::RHOQ_V] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VX_V] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VY_V] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::VZ_V] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_11_V] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_22_V] = 0.0;
+               mpiGrid[cellID]->parameters[CellParams::P_33_V] = 0.0;
+               for (uint popop = 0; popop < getObjectWrapper().particleSpecies.size(); ++popop) {
+                  Population & pop = mpiGrid[cellID]->get_population(popop);
```

Comment:
same for copysphere:
you cannot calculate velocities like this, you'll need to sum up momentum and then divide by sum of mass

Reply:
Maybe actually not needed, WIP